### DISCUSSION TITLE
Task Concepts

### DISCUSSION THREAD

**Discussion by malt-r** [Programmiermethoden/Dungeon#6] [Ideas]
This discussion holds concepts for tasks in the dungeon, which will be used as guides for evaluating requirements for the framework.

**Comment by malt-r** [comment]
# DFA Treasurehunt

## Aufgabentext

Beim Betreten des ersten Dungeon-Raums erscheint ein "Kartenitem" auf dem Boden, 
beim Aufheben erscheint folgender Text:
"Beim Erkunden des Dungeons hast du eine alte Zeichnung gefunden, die den Titel
'DFA' trägt. Hierbei scheint es sich um eine Art Schatzkarte zu handeln. Aber die 
Karte ist unvollständig. Damit sie gelesen werden kann benötigst du noch weitere 
Informationen. Sieh dich weiter im Dungeon um..." 

Beim Betreten des nächsten Raums erscheint eine Zeichenfolge, die den Eingabewörtern 
des DFA entspricht.
Text: "Die erscheinenen Zeichen ähneln den Zeichen an den Pfeilen der Schatzkarte,
vielleicht lässt sich so ein Weg durch die Räume des Dungeons finden."

## Ablauf

![dfa_treasurehunt_flowchart](https://user-images.githubusercontent.com/37513253/191435115-8ae438e2-2aa0-4844-9643-ecb445191f81.png)

## Was machen die Studierenden

"Suchen" einen Weg durch Räume im Dungeon mithilfe der Zustandsübergänge eines DFA.
Die Räume entsprechen dabei den Zuständen im DFA. In jedem Raum können Türen für die 
möglichen Folgezustände betreten werden und ein Zeichen (die aktuelle Eingabe für 
den DFA) ist abgebildet. 

Betritt der/die Studierende einen Raum, welcher mit aktuellem
Zustand und aktueller Eingabe dem nächsten DFA-Zustand entspricht, kommt er/sie
in einen Raum, der wie der vorherige, Türen für Zustandsübergänge und ein Zeichen enthält.
Falls der Raum nicht dem nächsten DFA-Zustand entspricht, muss der/die Studierende 
gegen ein Monster kämpfen und wieder den Weg zurück nehmen.

Gelangt der/die Studierende korrekt ans Ende dieser Dungeon-Räume, bekommt er eine 
Belohnung aus einer Schatzkiste.

## Was lernen die Studierenden

Funktion eines DFA.

## Welche Features werden benötigt?

- Anzeige von Bild (DFA, evtl. togglebar)
  - Auch evtl. anzeigen, in welchem Zustand wir gerade sind
- Questtext anzeigen
- kleine Dungeons mit Türen, Raumübergängen und "Türschildern" generieren
  - immer abhängig davon, welche Tür zuvor betreten wurde und in welchem Zustand 
  sich der DFA gerade befindet
  - Zusammen mit DFA und Eingabewort könnte das automatisiert ablaufen
- Inputsymbole für DFA anzeigen
- Schatztruhe (Inventarsystem)
- Kampfsystem (gegen Monster kämpen als Bestrafung bei betreten von falschem Raum)
- Verdecken von nicht betretenen Räumen
- Tracking von Spielerbewegung in durch die Räume (gewählte Türen)

## Wie wird die Aufgabe bewertet?

Verhältnis von richtig gewählten Türen beim ersten Versuch zu allen richtigen Türen.

## Wie sieht die Eingabe der Lehrperson aus (grobgranular)?

- Definition der Questtexte
- Generierung von Räumen & Inputzeichen für DFA muss definiert werden, so dass 
  die Aufgabe lösbar ist (idealerweise keine festkodierte Lösung)
  - Alternativ: Pool von möglichen DFAs definieren
- Transitionen zwischen Räumen definieren
- Was ist in einem Raum enthalten
- Interaktion mit Spielobjekten
  - Aufheben von Karteniten -> Aktivierung des Questexts
- DFA muss generiert, eingebunden und dargestellt werden

**Reply by malt-r** [reply]
# Technische Details DFA Treasurehunt

## Notizen, offene Punkte

- Räume geometrisch verbinden vs. Teleporter zwischen Räumen: siehe [hier](https://github.com/Programmiermethoden/Dungeon/discussions/19#discussioncomment-3813802) 
  
## technischer Ablauf

### Setup

- Generierung von Startraum & isolierten Räumen mit Türen/Teleportern & Schildern basierend 
  auf DFA
  - zu klären: soll einfach ein DFA festgelegt werden, der dann vom Interpreter 
  mehr oder weniger direkt an den Generator weitergegeben wird oder soll in 
  der DSL eine Generator-Funktion geschrieben werden, die einzelne Räume erstellt 
  und miteinander verbindet?
- Platzierung von Spieler und Kartenitem im Startraum/-areal
- Platzierung von Teleporter-"Türen" im Startraum/-areal

### Spiel

- Spieler hebt beim Erkunden Kartenitem auf (Interaktion)
- Fenster mit Aufgabentext öffnet sich:
  "Du hast eine Schatzkarte gefunden, die den Titel "DFA" trägt. Allerdings 
  scheint sie unvollständig zu sein. Sieh dich weiter im Dungeon um."
  - Klick auf Weiter -> Erklärung: "Mit m (Beispiel) kannst du dir die Schatzkarte ansehen" -> Klick auf weiter schließt das Fenster
  - Taste m: Toggle von Fenster in oberer Bildschirmecke, in dem DFA abgebildet ist
- nach weiterem Erkunden des aktuellen Areals betritt Spieler einen Teleporter
  - Teleportation in nächstes Areal mit erstem Raum, der einem DFA-Zustand entspricht
- Spieler betritt einen Raum mit drei Teleportern (an denen "Schilder" angebracht sind) und 
  einer Schild mit einem Eingabesymbol (für den DFA)
  - Interaktion mit Schild -> Fenster mit Namen des entsprechenden nächsten Zustands 
    oder des aktuellen Eingabesymbols geht auf
- Spieler interagiert mit einem Teleporter
  - Logging von Aktion als Bewertungsgrundlage -> setzt voraus, dass Teleporter 
  als "richtig" und "falsch" gekennzeichnet sind
  - Spieler wählt Teleporter mit korrektem nächsten Zustand aus
    - Laden von nächstem Areal & Platzierung von Spieler an Startposition
  - Spieler wählt Teleporter mit falschem nächsten Zustand aus -> Kampfareal
    - Platzierung von Monstern im Kampfareal (kann mehrfach betreten werden, immer,
    wenn ein falscher Teleporter ausgewählt wird)
    - Textblase über Monstern (oder Dialogfenster): "Hah, falsch ausgewählt!"
    - Spieler muss gegen Monster kämpfen
    - nachdem alle Monster besiegt sind, öffnet sich ein Teleporter, der 
      wieder zum vorherigen Areal führt
- Spieler findet den Endraum (Raum, der dem Endzustand des DFA entspricht), durch 
  Durchführung der vorher beschriebenen Auswahl von Teleportern
  - Spieler kann eine Truhe öffnen und bekommt eine Belohnung (Item)
  - Spieler nimmt nächsten Teleporter und beendet so das Level
- Basierend auf den gewählten Teleportern (richtig/falsch) kann eine Bewertung 
  erstellt werden

**Comment by malt-r** [comment]
Diskussion am 15.09.22 zu Aufgabentypen: es sollten auch recht einfache Aufgabentypen realisierbar sein, z.B.:
- Token im Raum in bestimmter Reihenfolge aufsammeln, ergibt am Ende ein Passwort
- Single Choice
- Multiple Choice
- Lückentexte
- Dinge aufsammeln und sortieren
- Monster greift mit einem String an, um es zu besiegen, muss die passende regex eingegeben werden

**Comment by AlexJ2017** [comment]
# Bäume und Graphen
## Aufgabentext
“Erkunde den Dungeon sorgsam, öffne mit Hilfe der Hinweise die Schatztruhe.”
Sobald der Spieler mit der Truhe interagiert, bekommt er einen Hinweis:
"Auf der Truhe steht etwas geschrieben: _Das Codewort lautet: Pre-Order_"
![Dungeon_Tree_01](https://user-images.githubusercontent.com/29783124/191743130-35cd58f3-6c5d-43f8-974a-0b7656b1de2e.jpg)
## Ablauf
Der Spieler startet in einem Raum mit einer verschlossenen Truhe. Jeder Raum im Dungeon hat eine Nummer zugeordnet, die der Spieler erst sieht, wenn er den Raum betritt. Die Truhe hat ein Zahlenschloss, das sich nur mit der richtigen Aneinanderreihung der Raumnummern öffnen lässt. Der Dungeon ist aufgebaut wie ein (Binär-)Baum. Je nach Aufgabenvariante muss dieser in Pre-, In- oder Post-Order Traversierungsreihenfolge als Code für die Schatztruhe eingegeben werden. Der Spieler kann sich frei in dem Dungeon bewegen und die Räume erkunden. Wenn er sich sicher ist, dass er die richtige Lösung hat, kann er diese an der Truhe austesten.
## Was lernen die Studierenden
Der Student lernt am Beispiel des Dungeons zu abstrahieren und eine Baumstruktur zu erstellen. Er lernt die verschiedenen Arten der Baum-Traversierung anzuwenden.
## Welche Features werden benötigt?
- Dungeon Generierung durch vorgegebenen (Binär-)Baum
    - Evtl. nicht-euklidische Raumanordnung (bei Platzmangel)
- Texteingabe (Truhen-Code)
## Wie wird die Aufgabe bewertet?
- Anzahl der Versuche zum Öffnen der Truhe
## Wie sieht die Eingabe der Lehrperson aus (grobgranular)?
- Definieren des Aufgaben Textes
- Bestimmen des Baums oder Graphen: 11(5(2, 8(6, 9)), 19(13, 30(-, 35)))
- Auswahl über Pre-, In- oder Postorder Aufgabentyp

**Comment by AMatutat** [comment]
# Statemachine

## Aufgabentext

Ihr Roboter-Held befindet sich im Kampf gegen einen starken Monster-Roboter. Sie haben es geschafft, sich Zugriff auf die Verhaltenssteuerung des Monsters verschafft. Helfen Sie ihrem Roboter, indem Sie die Reaktion des Monsters auf die Angriffe ihres Roboters umprogrammieren. 

*Erklärung: Die Zustände der Statemachine sind die Reaktionen des Monsters. Denkbar wären Wütend, Betäubt, Verwirrt, Angreifend etc.*

Ihr Roboter greif in folgender Reihenfolge an: 
Schwert, Schwert, Zauber, Zauber, Bogen, Schwert, Zauber, Schwert, Schwert. (*Erklärung: Das sind die Übergänge in der Statemachine*)

*Anmerkung: Hier sollte man weg von einer festen Eingabe und hin zu etwas in Richtung "Der Automat akzeptiert Wörter, die ein Palindrom sind" o.ä* 

## Ablauf

![Statemachine Boss](https://user-images.githubusercontent.com/32962412/194323953-e8d10b5b-5a21-43dc-a2ff-e7da237f98ae.png)

Der Spieler muss mithilfe der Bausteine (unten rechts in grün) bestehend aus Zuständen und Übergängen eine Statemachine konzeptionieren, welche der Anforderung entspricht (oben rechts). Dazu zieht er mithilfe von Drag&Drop die Bausteine in das dafür vorgesehene Feld (mitte rechts). 

Ist die Statemachine fertig konzeptioniert, wird der Held das Monster automatisch angreifen, dabei verwendet er eine Angriffsreihenfolge, die der Automat versteht. Ist der Automat richtig konzeptioniert, wird das Monster besiegt, wenn nicht, wird der Held verlieren. 




## Eingabe Lehrperson
- Definition welche Eingaben der Automat akzeptieren soll. 
- ggf. maximale Anzahl an Zuständen/Übergängen

## Bewertung
Bewertet wird zuerst die Korrektheit des Automaten. Erfüllt dieser die Anforderung? 
Zweites Bewertungskriterium ist die Anzahl der verwendeten Zustände und Übergänge. Wurden unnötige Übergänge/Zustände verwendet?  

## Lernziel
Konzeptionieren eigener Automaten welche vorher definierte Eingaben/Anforderungen akzeptieren. 

## Benötigte Features
- Boss Monster
- Statemachine
- Drag and Drop Elemente
- "Zeichnung"->Logik Schnittstelle

**Comment by AMatutat** [comment]
# Lückentext, Single und Multiple Choice fragen

## Aufgabentext
Hallo Held, kannst du mir diese Frage beantworten:
"INSERT TOLLE FRAGE"

## Ablauf
Der Spieler begnet einen NPC der ihm eine Frage stellt. 
Per Auswahl/Ankreuzen/Texteingabe kann die Frage beantwortet werden. Liegt der Spieler richtig, wird er mit einem Item belohnt.  

## Eingabe Lehrperson
- Fragestellung
- Richtige Antwort(en)
- Falsche Antworten

## Bewertung
Bewertet wird, ob die Frage richtig oder falsch Beantwortet wurde. 

## Lernziel
Beantworten verschiedener Theoretischer Fragestellung.

## Benötigte Features
- NPC
- Dialogfenster
- Auswahlmöglichkeiten (SC;MC)
- Texteingabemöglichkeit
  - Auswerten der Eingabe

**Reply by AMatutat** [reply]
Nach Rücksprache mit SHK, Lückentext (one liner) sollte gut zu gehen.
Multiline scheint schwer zu implementieren zu sein und es fehlen grundsätzliche Funktionen wie Scrolling etc.

**Reply by cagix** [reply]
> Nach Rücksprache mit SHK, Lückentext (one liner) sollte gut zu gehen.
> 
> Multiline scheint schwer zu implementieren zu sein und es fehlen grundsätzliche Funktionen wie Scrolling etc. 

Wie wäre es mit etwas Spiel-näheren Mechaniken: Der Spieler könnte Items droppen (richtige Items, richtige Reihenfolge, richtige Anzahl)? Der Spieler könnte mehrere vorhandene Kisten aufmachen (richtige Kisten, richtige Reihenfolge, ...)? Das passt besser zum Spiel und wir können bei Text-Einzeilern bleiben.

Das Übersetzen der Aufgabe in ein konkretes Szenario (Texteingabe oder Items droppen oder Kisten aufmachen) wäre für mich Sache der DSL - hier könnte man dann auch steuern, dass nicht immer die selbe Mechanik genutzt wird und dass die passenden Items irgendwo zu finden sind.

@bcg7 Was denkst Du dazu?

**Comment by malt-r** [comment]
# Escape-Room

Inspiriert von [einem heise Artikel](https://www.heise.de/ratgeber/Wie-Sie-mit-PowerPoint-ein-Escape-Spiel-erzeugen-7162654.html?view=print)

Könnte im Dungeon auf unterschiedliche Weise umgesetzt werden, bspw. durch verschiedene Räume, in denen Aufgaben / Rätsel / Fragen richtig beantwortet werden müssen, um Schlüsselfragmente zu bekommen, mit denen anschließend die Ausgangstür aus dem Level oder eine Truhe aufgeschlossen werden kann.

**Comment by bcg7** [comment]
Ich finde, unsere Aufgaben sollten möglichst keine (fast) direkt Übersetzung von Papieraufgaben sein und, wenn möglich, auch nicht deren "Denke" implementieren. Da fehlen uns sicher noch ein paar gute Ideen.

**Reply by AMatutat** [reply]
Ich bin da irgendwie gespalten. Auf der einen Seite stimme ich dir voll zu, wir haben ein "neues" Medium und damit kommen neue Anforderungen an die Aufgabenstellungen.
Auf der anderen Seite soll das am Ende auch genutzt werden und wenn ich mir erst eine neue Art Aufgaben zu denken beibringen muss, dann bleib ich um Stress doch wieder bei Zettel und Papier. 


Eigentlich wäre es perfekt klassische Aufgaben in das Ding einzugeben und dann passiert schwarze Magie und verwandelt das in ganz neue innovative Aufgaben.

**Reply by cagix** [reply]
Ich glaube, die Basissachen sollten zumindest unterstützt werden, damit die Transition einfacher wird. Man wird zuerst schauen, möglichst viele der alten Aufgaben zu übernehmen, und dann erst später wird man anfangen, sich schicke neue Aufgaben auszudenken, die dem Medium dann auch gerechter werden. Also ich sehe beide Seiten :)

**Reply by malt-r** [reply]
@cagix ich sehe das auch so, da die Erstellung neuer, kreativer Aufgabenformate ja auch ein gewisses Vertrautsein mit der DSL und dem Gesamtsystem voraussetzt. Die angesprochene Übergangszeit sollte da in jedem Fall mitgedacht werden, denke ich
